numpyro/mcmc.py

```diff
@@ -428,3 +441,303 @@ def single_chain_mcmc(rng, init_params):
+                                   step_size=step_size, adapt_step_size=adapt_step_size,
+                                   adapt_mass_matrix=adapt_mass_matrix, dense_mass=dense_mass,
+                                   target_accept_prob=target_accept_prob, trajectory_length=trajectory_length)
+        self.max_tree_depth = max_tree_depth
+        self.algo = 'NUTS'
+
+
+class MCMC(object):
+    """
+    Provides access to Markov Chain Monte Carlo inference algorithms in NumPyro.
+
+    :param MCMCKernel sampler: an instance of :class:`~numpyro.mcmc.MCMCKernel` that
+        determines the sampler for running MCMC. Currently, only :class:`~numpyro.mcmc.HMC`
+        and :class:`~numpyro.mcmc.NUTS` are available.
+    :param int num_warmup: Number of warmup steps.
+    :param int num_samples: Number of samples to generate from the Markov chain.
+    :param int num_chains: Number of Number of MCMC chains to run. By default,
+        chains will be run in parallel using :func:`jax.pmap`, failing which,
+        chains will be run in sequence.
+    :param constrain_fn: Callable that converts a collection of unconstrained
+        sample values returned from the sampler to constrained values that
+        lie within the support of the sample sites.
+    :param bool progress_bar: Whether to enable progress bar updates. Defaults to
+        ``True``.
+    """
+    def __init__(self,
+                 sampler,
+                 num_warmup,
+                 num_samples,
+                 num_chains=1,
+                 constrain_fn=None,
+                 progress_bar=True):
+        self.sampler = sampler
+        self.num_warmup = num_warmup
+        self.num_samples = num_samples
+        self.num_chains = num_chains
+        self.constrain_fn = constrain_fn
+        self.progress_bar = progress_bar
+        self.sequential_chain = False
+        if xla_bridge.device_count() < num_chains:
+            self.sequential_chain = True
+            warnings.warn('There are not enough devices to run parallel chains: expected {} but got {}.'
+                          ' Chains will be drawn sequentially. If you are running `mcmc` in CPU,'
+                          ' consider to disable XLA intra-op parallelism by setting the environment'
+                          ' flag "XLA_FLAGS=--xla_force_host_platform_device_count={}".'
+                          .format(num_chains, xla_bridge.device_count(), num_chains))
+        # TODO: We should have progress bars (maybe without diagnostics) for num_chains > 1
+        if num_chains > 1 or "CI" in os.environ or "PYTEST_XDIST_WORKER" in os.environ:
+            self.progress_bar = False
+
+        self._samples = None
+        self._samples_flat = None
+
+    def _single_chain_mcmc(self, init, args=(), kwargs={}):
+        init_params = None
+        if len(init) == 1:
+            rng, = init
+        else:
+            rng, init_params = init
```

Comment:
Hmm, `pmap`, `lax.map` supports batching over `None` and it is clear that if `init_params` is not specified, then until here it is still None. I think we can simplify this with
```
rng, init_params = init

# and at `run` method
samples = map_fn((rngs, init_params))
```

Reply:
Oh..I didn't realize batching over None was allowed. Let me simplify that.